kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java

```diff
@@ -233,6 +239,58 @@ void shouldDecryptEncryptedRecords() {
         ));
     }
 
+    @Test
+    void fetchShouldMaintainBrokerOffsets() {
+        // Given
+        var recsWithNonConsecutiveOffsets = makeRecords(IntStream.of(0, 2, 3), (u) -> new TestingRecord(ByteBuffer.wrap(HELLO_CIPHER_WORLD)));
+        var encryptedFetchResponse = buildFetchResponseData(new FetchableTopicResponse()
+                .setTopic(ENCRYPTED_TOPIC)
+                .setPartitions(List.of(new PartitionData().setRecords(recsWithNonConsecutiveOffsets))));
+
+        when(keyManager.decrypt(any(), anyInt(), assertArg(records -> assertThat(records).hasSize(3)), any(Receiver.class))).thenAnswer(invocationOnMock -> {
+            final List<Record> records = invocationOnMock.getArgument(2);
+            final Receiver receiver = invocationOnMock.getArgument(3);
+
+            // Oh my!!
+            records.forEach(r -> {
+                receiver.accept(r, ByteBuffer.wrap(HELLO_PLAIN_WORLD), Record.EMPTY_HEADERS);
+            });
+            return CompletableFuture.completedFuture(null);
+        });
+
+        // When
+        encryptionFilter.onFetchResponse(ProduceResponseData.HIGHEST_SUPPORTED_VERSION,
+                new ResponseHeaderData(), encryptedFetchResponse, context);
+
+        // Then
+        verify(context, times(1)).forwardResponse(any(ResponseHeaderData.class), assertArg(actualFetchResponse -> assertThat(actualFetchResponse)
+                        .isInstanceOf(FetchResponseData.class)
+                        .asInstanceOf(InstanceOfAssertFactories.type(FetchResponseData.class))
+                        .has(new FetchResponseDataCondition(fetchResponseData -> {
+
+//                            System.out.println("frd");
+//                            fetchResponseData.responses().forEach(r -> {
+//                                System.out.println("r " + r);
+//                                r.partitions().forEach(p -> {
+//                                    System.out.println("p" + p);
+//                                });
+//
+//                            });
+
+                            // Yikes!
+                            var offsets = fetchResponseData.responses().stream()
+                                    .flatMap(pd -> pd.partitions().stream())
+                                    .map(PartitionData::records)
+                                    .map(org.apache.kafka.common.record.Records.class::cast)
+                                    .flatMap(r -> StreamSupport.stream(r.records().spliterator(), false))
```

Comment:
This feels very similar to something I was poking at recently. I wonder about pulling this out a util function returning a `Stream<kafka.Record>`